### PR TITLE
Implement Progress bar Helper

### PR DESCRIPTION
**Pull request by markstory**
This adds a progress bar shell helper. I've used it in the `i18n extract` task as it seemed like a natural fit there. It provides two usage scenarios:

``` php
// simple use
$func = function ($progress) {
    sleep(1);
    $progress->increment(rand(5, 20));
};
$this->_io->progress(['callback' => $func]);

// manual usage
$progress = $this->_io->helper('progress');
// Progress through 300 things
$progress->init(['total' => 300]);
$progress->increment(20);
$progress->draw();
```

I've also cleaned up the ConsoleIo class a bit and making testing stubs better.

### CODE REVIEW

File: tests/TestCase/Shell/Helper/ProgressHelperTest.php [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+     *
+     * @expectedException \RuntimeException
+     */
+    public function testOutputFailure()
+    {
+        $this->helper->output(['not a callback']);
+    }
+
+    /**
+     * Test a callback that never reaches 100 fails.
```
Comment (antograssiot) [inline tests/TestCase/Shell/Helper/ProgressHelperTest.php]: Not sure that this is what the following test is doing

File: src/Shell/Helper/ProgressHelper.php [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+/**
+ * CakePHP(tm) : Rapid Development Framework (http://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @link          http://book.cakephp.org/2.0/en/development/testing.html CakePHP(tm) Tests
```
Comment (antograssiot) [inline src/Shell/Helper/ProgressHelper.php]: this link seems outdated

**Comment by antograssiot** [comment]
Looks great :+1:, also you have some [CS issue](https://travis-ci.org/cakephp/cakephp/jobs/6373520)

**Comment by markstory** [comment]
@antograssiot Fixed up now.

**Comment by ADmad** [comment]
PHPCS build still failing

**Comment by markstory** [comment]
@ADmad Would have been nice if phpcs gave all the errors at once :cry:

**Comment by ADmad** [comment]
It did show it [earlier](https://travis-ci.org/cakephp/cakephp/jobs/63735089#L229) too :smile:

**Comment by markstory** [comment]
Well then, I just cannot read very well.

**Comment by markstory** [comment]
Any other feedback on this or should I just merge it?

**Comment by dakota** [comment]
:+1: from me, I reckon good to merge.

**Comment by josbeir** [comment]
Not entirely sure but does it supposed to look like this ? (in iTerm but also in standard mac terminal)

Also experiencing this with the i18n shell for example :-)

screen shot 2015-10-16 at 14 47 06

**Comment by markstory** [comment]
@josbeir No. That's not how its supposed to look.  Interestingly, it looks fine for me on OSX in iTerm.

**Comment by josbeir** [comment]
Very weird, just tried it on another (osx 10.11) pc, created a new cake 3.1.1 project and created a shell script using the following code

```php
    public function main()
    {
        $total = 50;
        $progress = $this->helper('progress');
        $progress->init([
            'total' => $total
        ]);

        for ($x = 0; $x <= $total; $x++) {
            $progress->increment($x);
            $progress->draw();
            sleep(1);
        }
    }
```

Still experiencing the issue (screenshot)

screen shot 2015-10-17 at 13 11 27

**Comment by HavokInspiration** [comment]
I tried your code on my ubuntu workstation and I have a similar issue.
I might have found a solution, I need to run more tests.

**Comment by antograssiot** [comment]
The solution is easy, I'll open a PR
just need a ceil() to handle float on pad string

**Comment by markstory** [comment]
@josbeir Ok, I see a similar results with your example shell.

**Pull request merged**